File: ocra/analysis/analysis_result.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Dict, Any

from ocra.models.pose_frame import PoseFrame
from ocra.biomechanics.biomechanical_frame import BiomechanicalFrame


@dataclass
class AnalysisResult:
    """
    Container principal del análisis biomecánico.

    Contiene:

    - frames originales de Kinovea
    - frames biomecánicos calculados
    - resultados temporales de postura

    No contiene:
    - puntuación OCRA
    - evaluación normativa
    - nivel de riesgo
    """

    pose_frames: List[PoseFrame]

    biomechanical_frames: List[BiomechanicalFrame]

    posture_results: Dict[str, Any] = field(
        default_factory=dict
    )

    metadata: Dict[str, Any] = field(
        default_factory=dict
    )


    def as_dict(self) -> Dict[str, Any]:
        """
        Resumen exportable.
        """

        return {

            "metadata": dict(self.metadata),

            "num_pose_frames":
                len(self.pose_frames),

            "num_biomechanical_frames":
                len(self.biomechanical_frames),

            "posture_results":
                {
                    key:
                    value.as_dict()
                    if hasattr(value, "as_dict")
                    else value

                    for key, value
                    in self.posture_results.items()
                }
        }

File: js/analysis_result.js
"use strict";

/*
=========================================================
Analysis Result Model

Equivalente al AnalysisResult de Python.

Responsabilidades:

- Mantener el resultado completo del análisis.
- Separar datos originales de datos calculados.
- No contiene evaluación OCRA.
- No contiene nivel de riesgo.

Contiene:

- frames originales Kinovea
- frames anatómicos adaptados
- resultados biomecánicos
- resultados postura
- resultados movimientos
- metadatos

=========================================================
*/


class AnalysisResult {


    constructor() {


        this.metadata = {};


        /*
        -----------------------------------------
        Datos originales
        -----------------------------------------
        */

        this.poseFrames = [];



        /*
        -----------------------------------------
        Datos preparados para biomecánica
        -----------------------------------------
        */

        this.anatomicalFrames = [];



        /*
        -----------------------------------------
        Resultados biomecánicos
        -----------------------------------------
        */

        this.biomechanicalFrames = [];



        /*
        -----------------------------------------
        Resultados temporales
        -----------------------------------------
        */

        this.postureResults = {};

        this.movementResults = {};



        /*
        -----------------------------------------
        Evaluación OCRA

        Se mantiene vacío.
        El cálculo lo realizará el especialista.
        -----------------------------------------
        */

        this.ocra = null;


    }



    /*
    =====================================================
    Cargar metadatos
    =====================================================
    */


    setMetadata(
        data
    ) {


        this.metadata = {

            ...this.metadata,

            ...data

        };


    }





    /*
    =====================================================
    Añadir frames Kinovea
    =====================================================
    */


    setPoseFrames(
        frames
    ) {


        this.poseFrames =
            Array.isArray(frames)
            ?
            frames
            :
            [];


    }





    /*
    =====================================================
    Añadir frames anatómicos
    =====================================================
    */


    setAnatomicalFrames(
        frames
    ) {


        this.anatomicalFrames =
            Array.isArray(frames)
            ?
            frames
            :
            [];


    }





    /*
    =====================================================
    Añadir resultados biomecánicos
    =====================================================
    */


    setBiomechanicalFrames(
        results
    ) {


        this.biomechanicalFrames =
            Array.isArray(results)
            ?
            results
            :
            [];


    }





    /*
    =====================================================
    Exportación simple
    =====================================================
    */


    asObject() {


        return {


            metadata:
                this.metadata,


            numPoseFrames:
                this.poseFrames.length,


            numAnatomicalFrames:
                this.anatomicalFrames.length,


            numBiomechanicalFrames:
                this.biomechanicalFrames.length,



            postureResults:
                this.postureResults,


            movementResults:
                this.movementResults,


            ocra:
                this.ocra


        };


    }



}
